pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java

```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
+    //    the wire for intermediate aggregation.
+    if (childRel instanceof Project) {
+      // Avoid creating a new LogicalProject if the child node of aggregation is already a project node.
+      project = (LogicalProject) childRel;
+      newAggArgColumns = fetchNewAggArgCols(oldAggRel.getAggCallList());
+      newAggGroupByColumns = oldAggRel.getGroupSet().asList();
+    } else {
+      // Create a leaf stage project. This is done so that only the required columns are sent over the wire for
+      // intermediate aggregation. If there are multiple aggregations on the same column, the column is projected
+      // only once.
+      project = createLogicalProjectForAggregate(oldAggRel, newAggArgColumns, newAggGroupByColumns);
+    }
+
+    // 2. Create an exchange on top of the LogicalProject.
+    LogicalExchange exchange = LogicalExchange.create(project, RelDistributions.hash(newAggGroupByColumns));
+
+    // 3. Create an intermediate stage aggregation.
+    RelNode newAggNode =
+        makeNewIntermediateAgg(call, oldAggRel, exchange, false, newAggArgColumns, newAggGroupByColumns);
+
+    call.transformTo(newAggNode);
+  }
+
+  private LogicalProject createLogicalProjectForAggregate(Aggregate oldAggRel, List<Integer> newAggArgColumns,
+      List<Integer> newAggGroupByCols) {
+    RelNode childRel = ((HepRelVertex) oldAggRel.getInput()).getCurrentRel();
+    RexBuilder childRexBuilder = childRel.getCluster().getRexBuilder();
+    List<RelDataTypeField> fieldList = childRel.getRowType().getFieldList();
+
+    List<RexNode> projectColRexNodes = new ArrayList<>();
+    List<String> projectColNames = new ArrayList<>();
+    // Maintains a mapping from the column to the corresponding index in projectColRexNodes.
+    Map<Integer, Integer> projectSet = new HashMap<>();
+
+    int projectIndex = 0;
+    for (int group : oldAggRel.getGroupSet().asSet()) {
+      projectColNames.add(fieldList.get(group).getName());
+      projectColRexNodes.add(childRexBuilder.makeInputRef(childRel, group));
+      projectSet.put(group, projectColRexNodes.size() - 1);
+      newAggGroupByCols.add(projectIndex++);
+    }
+
+    List<AggregateCall> oldAggCallList = oldAggRel.getAggCallList();
+    for (int i = 0; i < oldAggCallList.size(); i++) {
+      List<Integer> argList = oldAggCallList.get(i).getArgList();
+      if (argList.size() == 0) {
+        newAggArgColumns.add(-1);
+        continue;
+      }
+      for (int j = 0; j < argList.size(); j++) {
+        Integer col = argList.get(j);
+        if (!projectSet.containsKey(col)) {
+          projectColRexNodes.add(childRexBuilder.makeInputRef(childRel, col));
+          projectColNames.add(fieldList.get(col).getName());
+          projectSet.put(col, projectColRexNodes.size() - 1);
+          newAggArgColumns.add(projectColRexNodes.size() - 1);
+        } else {
+          newAggArgColumns.add(projectSet.get(col));
```

Comment:
I think it is possible to do `newAggArgColumns.add()` commonly outside the if-else block.

(nit) nbd.. but just seeing how we can make this code slightly more readable / less verbose :)